src/docs/user/ProActiveUserGuide.adoc

```diff
@@ -1763,6 +1710,234 @@ println outFile.text
+[29t0@trydev.activeeon.com;14:43:49] Sleeping 5 seconds
+[29t1@trydev.activeeon.com;14:43:48] Waiting for Task A
+[29t0@trydev.activeeon.com;14:43:54] Unlocking Task B
+[29t1@trydev.activeeon.com;14:43:55] Task B has been unlocked by Task A
+[29t3@trydev.activeeon.com;14:44:01] Channel 29 deleted.
+```
+
+The Complete Workflow example can be downloaded link:examples/KeyValueWait.xml[here].
+
+==== Example 3 : Critical Section
+
+In this last example, we will show how we can create a *critical section* inside a workflow, which means, a section of the workflow which can only be accessed by a configurable number of *slots*.
+
+The _Critical Section_ workflow is defined with a <<_replicate, Replicate Control Structure>>:
+
+image::CriticalSection.png[Critical Section, , title="Critical Section"]
+
+It contains the following Tasks:
+
+ * *Init*: Initialize the channel. This task creates a channel using the current _job id_. It also sets a `available_slots` binding in the key/value store with an initial value based on the workflow user-defined `NB_SLOTS` <<_job_variables, job variable>>.
+```groovy
+jobid = variables.get("PA_JOB_ID")
+synchronizationapi.createChannel(jobid, false)
+synchronizationapi.put(jobid, "available_slots", Integer.parseInt(variables.get("NB_SLOTS")))
+println "Channel " + jobid + " created."
+```
+This tasks also define a <<_control_flow_scripts, control flow script>> used to define the number of replicated tasks. The number of replicated tasks is parametrized with the user-defined `NB_TASKS` <<_job_variables, job variable>>:
+```groovy
+runs=Integer.parseInt(variables.get("NB_TASKS"))
+```
+
+ * *Section*: Wait for the critical section to be available. This task will block until the number of available slots is greater than 0. As soon as at least one slot is available, it decrements the `available_slots` counter and perform some computation.
+When the computation is terminated, it releases the slot by incrementing back the `available_slots` counter.
+```groovy
+println "Waiting for available slot"
+jobid = variables.get("PA_JOB_ID")
+synchronizationapi.waitUntilThen(jobid, "available_slots", "{k, x -> x > 0}", "{ k, x -> x - 1}")
+println "Enter critical section"
+// simulate some computation
+Thread.sleep(10000)
+println "Release slot"
+synchronizationapi.compute(jobid,"available_slots", "{ k, x -> x + 1}")
+```
+
+As you can see, the first method used in this section is https://www.activeeon.com/public_content/documentation/javadoc/dev/org/ow2/proactive/scheduler/synchronization/Synchronization.html#waitUntilThen-java.lang.String-java.lang.String-java.lang.String-java.lang.String-[waitUntilThen]. This method expects both a https://docs.oracle.com/javase/8/docs/api/java/util/function/BiPredicate.html?is-external=true[BiPredicate] and a https://docs.oracle.com/javase/8/docs/api/java/util/function/BiFunction.html?is-external=true[BiFunction] defined as strings http://groovy-lang.org/closures.html[Groovy Closure].
```

Comment:
We can say `and a BiFunction to define what to do when the predicate is met. Both the BiPredicate and the BiFunction must be defined as strings Groovy Closure.`

Reply:
Ok this is said in the next sentence.